packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,477 @@
+  target: PartialPrimitiveTargetDescriptor,
+): RecursiveArray<string> {
+  const components: RecursiveArray<string> = [];
+  if (target.modifiers != null) {
+    components.push(target.modifiers.map(modifierToSpokenForm));
+  }
+  if (target.mark != null) {
+    components.push(markToSpokenForm(target.mark));
+  }
+  return components;
+}
+
+function modifierToSpokenForm(modifier: Modifier): RecursiveArray<string> {
+  switch (modifier.type) {
+    case "cascading":
+    case "modifyIfUntyped":
+      throw new NoSpokenFormError(`Modifier '${modifier.type}'`);
+
+    case "containingScope":
+      return [scopeTypeToSpokenForm(modifier.scopeType)];
+
+    case "everyScope":
+      return [modifiers.everyScope, scopeTypeToSpokenForm(modifier.scopeType)];
+
+    case "extendThroughStartOf":
+    case "extendThroughEndOf": {
+      const type = modifiers[modifier.type];
+      return modifier.modifiers != null
+        ? [type, modifier.modifiers.map(modifierToSpokenForm)]
+        : [type];
+    }
+
+    case "relativeScope":
+      return modifier.offset === 0
+        ? relativeScopeInclusiveToSpokenForm(modifier)
+        : relativeScopeExclusiveToSpokenForm(modifier);
+
+    case "ordinalScope": {
+      const scope = scopeTypeToSpokenForm(modifier.scopeType);
+
+      if (modifier.length === 1) {
+        if (modifier.start === -1) {
+          return [modifiersExtra.last, scope];
+        }
+        if (modifier.start === 0) {
+          return [modifiersExtra.first, scope];
+        }
+        if (modifier.start < 0) {
+          return [
+            ordinalToSpokenForm(Math.abs(modifier.start)),
+            modifiersExtra.last,
+            scope,
+          ];
+        }
+        return [ordinalToSpokenForm(modifier.start + 1), scope];
+      }
+
+      const number = numberToSpokenForm(modifier.length);
+
+      if (modifier.start === 0) {
+        return [modifiersExtra.first, number, pluralize(scope)];
+      }
+      if (modifier.start === -modifier.length) {
+        return [modifiersExtra.last, number, pluralize(scope)];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
+    }
+
+    case "range": {
+      if (
+        modifier.anchor.type === "ordinalScope" &&
+        modifier.active.type === "ordinalScope" &&
+        modifier.anchor.length === 1 &&
+        modifier.active.length === 1 &&
+        modifier.anchor.scopeType.type === modifier.active.scopeType.type
+      ) {
+        const anchor =
+          modifier.anchor.start === -1
+            ? modifiersExtra.last
+            : ordinalToSpokenForm(modifier.anchor.start + 1);
+        const active = modifierToSpokenForm(modifier.active);
+        const connective = getRangeConnective(
+          modifier.excludeAnchor ?? false,
+          modifier.excludeActive ?? false,
+        );
+        return [anchor, connective, active];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
```

Comment:
and this one, how come not `NoSpokenFormError`?

Reply:
See above

Reply:
This one I'm slightly more on the fence about, but again I think mostly the problem is just that we don't currently have a spoken form for it. I don't think there's anything inherently wrong with this case, and we do support it extension-side. But again this one's a bit different because spoken form would prob be ambiguous and we may remove this construct in the future. But I'm still not sure I see the benefit of not just throwing `NoSpokenFormError`. I think I'd reserve `Error` for when there's truly something wrong / unexpected. But if you want to push the other direction for this specific case I could be convinced